store/tikv/client.go

```diff
@@ -172,6 +292,126 @@ func (a *connArray) Close() {
+	err     error
+}
+
+func (a *connArray) batchSendLoop(cfg config.TiKVClient) {
+	entries := make([]*batchCommandsEntry, 0, cfg.MaxBatchSize)
+	requests := make([]*tikvpb.BatchCommandsRequest_Request, 0, cfg.MaxBatchSize)
+	requestIDs := make([]uint64, 0, cfg.MaxBatchSize)
+
+	for {
+		metrics.TiKVPendingBatchRequests.Set(float64(len(a.batchCommandsCh)))
+
+		// Choose a connection by round-robbin.
+		next := atomic.AddUint32(&a.index, 1) % uint32(len(a.v))
+		batchCommandsClient := a.batchCommandsClients[next]
+
+		tikvTransportLayerLoad := atomic.LoadUint64(batchCommandsClient.tikvTransportLayerLoad)
+		inHeavyLoad := uint(tikvTransportLayerLoad) >= cfg.TiKVHeavyLoadToBatch // Need to wait.
+
+		entries = entries[:0]
+		requests = requests[:0]
+		requestIDs = requestIDs[:0]
+
+		// Block on the first element.
+		headEntry := <-a.batchCommandsCh
+		if headEntry == nil {
+			return
+		}
+		entries = append(entries, headEntry)
+
+		// This loop is for trying best to collect more requests.
+	Loop:
+		for {
+			select {
+			case entry := <-a.batchCommandsCh:
+				if entry == nil {
+					return
+				}
+				entries = append(entries, entry)
+				if len(entries) >= int(cfg.MaxBatchSize) {
+					break Loop
+				}
+			default:
+				break Loop
+			}
+		}
+
+		// If the target TiKV is overload, wait a while to collect more requests.
+		if len(entries) < int(cfg.MaxBatchSize) && inHeavyLoad && cfg.MaxBatchWaitTime > 0 {
+			waitStart := time.Now()
+			var waitEnd time.Time
+			after := time.After(cfg.MaxBatchWaitTime)
+			batchIsFilled := false
+		WaitLoop:
+			for {
+				if !batchIsFilled {
+					select {
+					case entry := <-a.batchCommandsCh:
+						if entry == nil {
+							return
+						}
+						entries = append(entries, entry)
+						if len(entries) >= int(cfg.BatchWaitSize) {
+							batchIsFilled = true
+						}
+					case now := <-after:
+						waitEnd = now
+						break WaitLoop
+					}
+				} else {
+					select {
+					case entry := <-a.batchCommandsCh:
+						if entry == nil {
+							return
+						}
+						entries = append(entries, entry)
+						if len(entries) >= int(cfg.MaxBatchSize) {
+							waitEnd = time.Now()
+							break WaitLoop
+						}
+					default:
+						waitEnd = time.Now()
+						break WaitLoop
+					}
+				}
+			}
+			metrics.TiKVBatchWaitDuration.Observe(float64(waitEnd.Sub(waitStart)))
+		}
+
+		for _, entry := range entries {
+			requests = append(requests, entry.req)
+		}
+
+		length := len(requests)
+		maxBatchID := atomic.AddUint64(&batchCommandsClient.idAlloc, uint64(length))
+		for i := 0; i < length; i++ {
+			requestID := uint64(i) + maxBatchID - uint64(length)
+			requestIDs = append(requestIDs, requestID)
+			batchCommandsClient.batched.Store(requestID, entries[i])
+		}
+
+		request := &tikvpb.BatchCommandsRequest{
+			Requests:   requests,
+			RequestIds: requestIDs,
+		}
+
+		batchCommandsClient.clientLock.Lock()
+		err := batchCommandsClient.client.Send(request)
+		batchCommandsClient.clientLock.Unlock()
```

Comment:
question not a review :laughing:  : this long hold lock is required due `failPendingRequests`?

Reply:
There are 2 purpose about `clientLock`: one for protecting `batched` in `failPendingRequest`, and the another is protecting the `client` won't be replaced in `RecvLoop`, which runs in other threads.